flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/handler/PulsarAdminInvocationHandler.java

```diff
@@ -0,0 +1,152 @@
+import static org.apache.flink.util.ExceptionUtils.findThrowable;
+
+/** A wrapper which wraps the {@link PulsarAdmin} with request retry and rate limit support. */
+public class PulsarAdminInvocationHandler implements InvocationHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(PulsarAdminInvocationHandler.class);
+
+    @SuppressWarnings("java:S3077")
+    private static volatile RateLimiter rateLimiter;
+
+    private final PulsarAdmin admin;
+    private final int retryTimes;
+    private final long waitMillis;
+    private final int requestRates;
+    private final Map<String, Object> handlers;
+
+    public PulsarAdminInvocationHandler(PulsarAdmin admin, PulsarConfiguration configuration) {
+        this.admin = admin;
+        this.retryTimes = configuration.get(PULSAR_ADMIN_REQUEST_RETRIES);
+        this.waitMillis = configuration.get(PULSAR_ADMIN_REQUEST_WAIT_MILLIS);
+        this.requestRates = configuration.get(PULSAR_ADMIN_REQUEST_RATES);
+        this.handlers = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+        Class<?> returnType = method.getReturnType();
+
+        // No need to proxy the void return type.
+        // The non-interface type is not able to proxy.
+        if (returnType.equals(Void.TYPE) || !returnType.isInterface()) {
+            return method.invoke(admin, args);
+        }
+
+        String methodName = method.getName();
+        if (handlers.containsKey(methodName)) {
+            return handlers.get(methodName);
+        }
+
+        Object handler =
+                Proxy.newProxyInstance(
+                        Thread.currentThread().getContextClassLoader(),
+                        new Class[] {returnType},
+                        new RequestHandler(method.invoke(admin, args)));
+        this.handlers.put(methodName, handler);
+
+        return handler;
+    }
+
+    /** A proxy handler with retry support for all the admin request. */
+    @SuppressWarnings({"java:S1193", "java:S1181", "java:S3776", "java:S112"})
+    private class RequestHandler implements InvocationHandler {
+
+        private final Object handler;
+
+        public RequestHandler(Object handler) {
+            this.handler = handler;
+        }
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            return doInvoke(method, args, retryTimes);
+        }
+
+        private Object doInvoke(Method method, Object[] args, int remainingTimes) throws Throwable {
+            while (true) {
+                // Make sure the request is allowed in the given rates.
+                requestRateLimit(requestRates);
+
+                try {
+                    return method.invoke(handler, args);
+                } catch (InvocationTargetException e) {
+                    Optional<PulsarAdminException> optional =
```

Comment:
Maybe `e.getTargetException()` is enough, do we really need to find the `PulsarAdminException` in more deeper level.

Reply:
Yep. You are right.